torch/quantization/convert_modules.py

```diff
@@ -0,0 +1,240 @@
+        super(AbstractDeQuant, self).__init__()
+
+    def forward(self, x):
+        return x
+
+def calculateQParams(MyModule):
+    r""" Calculates Quantizer parameters for activation tensors based on observer statistics
+    Updates the qparams parameter for each module with the quantizer params.
+
+    Args:
+        MyModule: Model with observer stats
+    """
+    if hasattr(MyModule, 'observer'):
+        if isinstance(MyModule.observer, Observer):
+            # Simple symmetric quantization over entire observed range
+            MyModule.qparams = MyModule.observer.calculate_q_params()
+    for Module in MyModule.children():
+        calculateQParams(Module)
+
+def quantize(module, qConfigDict, eval_fn, *eval_args):
+    r"""Given a float module and qConfigDict, convert it to a quantized module
+
+    First it will prepare the module to add observer and abstract quant
+    deqaunt modules, then it calls `eval_fn` which will run the calibration
+    step and observers will record the stats of the tensors, after that we will
+    call `convert` which will convert the module to a quantized module.
+    """
+    prepare(module, qConfigDict)
+    eval_fn(module, *eval_args)
+    convert(module)
+    return
+
+def convert(module):
+    r""" Utility function to traverse module tree and swap modules marked with
+         AbstractQuant and AbstractDeQuant with
+         quantized versions.
+    """
+    calculateQParams(module)
+    return convertToQuantized(module)
+
+def convertToQuantized(module):
+    r"""Convert a module with qparams to a quantized version of the module
+    """
+    module_swapped = swapModule(module)
+    if len(list(module.named_children())) == 0:
+        return module
+
+    reassign = {}
+    for name, mod in module.named_children():
+        new_mod = convertToQuantized(mod)
+        if new_mod is not mod:
+            reassign[name] = new_mod
+
+    for name, mod in reassign.items():
+        setattr(module_swapped, name, mod)
+
+    return module_swapped
+
+def swapModule(mod):
+    r""" Check if a module has a quantized counterpart and swap it.
+    """
+    new_mod = mod
+    if hasattr(mod, 'observer'):
+        if type(mod) == torch.nn.Linear:
+            # Obtain quantized weights and biases
+            observerConfig = mod.qConfig['weight']['config']
+            weight_observer = mod.qConfig['weight']['observer'](observerConfig)
+            quantized_weight, quantized_bias = quantizeWeightAndBias(
+                mod, weight_observer)
+            new_mod = nnq.Linear(mod.out_features, mod.in_features)
+            state = {
+                'weight': quantized_weight,
+                'bias': quantized_bias,
+                'output_scale': mod.qparams[0],
+                'output_zero_point': mod.qparams[1]
+            }
+            new_mod.set_state(state)
+        if type(mod) == torch.nn.Identity:
+            new_mod = nnq.Quantize(mod.qparams[0].item(), mod.qparams[1].item(), torch.quint8)
```

Comment:
can we standardize interface for nnq module inputs? For example - always just accept an original module. Then the constructor can poke into qparams, original weights, etc. And you don't need a fancy swapModule function - it can just take the dictionary of module remappings and operate generically

Reply:
We had originally started with that, but moved to this approach. The reasoning is that one can create a quantized module from scratch and come up with any scheme for creating quantized weights and calculating scale/zero-points and still be able to run it in pytorch.  Having the conversion coupled with the quantized module forces us to mandate one way of doing the quantization, which is too restrictive